backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/questionsubmission/QuestionSubmissionService.java

```diff
@@ -0,0 +1,278 @@
+        return new QuestionSubmissionDto(questionSubmission);
+    }
+
+    @Retryable(value = {SQLException.class}, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public ReviewDto createReview(ReviewDto reviewDto) {
+        checkIfConsistentReview(reviewDto);
+
+        QuestionSubmission questionSubmission = getQuestionSubmission(reviewDto.getQuestionSubmissionId());
+
+        User user = getUser(reviewDto.getUserId());
+
+        Review review = new Review(user, questionSubmission, reviewDto);
+
+        updateQuestionStatus(review.getStatus(), questionSubmission.getQuestion().getId());
+
+        reviewRepository.save(review);
+        return new ReviewDto(review);
+    }
+
+    @Retryable(value = {SQLException.class}, backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.REPEATABLE_READ)
+    public QuestionSubmissionDto updateQuestionSubmission(Integer questionSubmissionId, QuestionSubmissionDto questionSubmissionDto) {
+        QuestionSubmission questionSubmission = getQuestionSubmission(questionSubmissionId);
+
+        User user = getStudent(questionSubmissionDto.getUserId());
+
+        if(user.isStudent() && questionSubmission.getQuestion().getStatus() != Question.Status.IN_REVISION) {
+            throw new TutorException(CANNOT_EDIT_REVIEWED_QUESTION);
+        }
+
+        this.questionService.updateQuestion(questionSubmissionDto.getQuestion().getId(), questionSubmissionDto.getQuestion());
+        return new QuestionSubmissionDto(questionSubmission);
+    }
+
+    @Retryable(
+            value = { SQLException.class },
+            backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    public void updateQuestionSubmissionTopics(Integer questionSubmissionId, TopicDto[] topics) {
+        Integer questionId = questionSubmissionRepository.findQuestionIdByQuestionSubmissionId(questionSubmissionId);
+
+        questionService.updateQuestionTopics(questionId, topics);
+    }
+
+    @Retryable(
+            value = { SQLException.class },
+            backoff = @Backoff(delay = 5000))
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    public void removeSubmittedQuestion(Integer questionSubmissionId) {
+        QuestionSubmission questionSubmission = questionSubmissionRepository.findById(questionSubmissionId).orElseThrow(() -> new TutorException(QUESTION_SUBMISSION_NOT_FOUND, questionSubmissionId));
+        Question question = questionRepository.findById(questionSubmission.getQuestion().getId()).orElseThrow(() -> new TutorException(QUESTION_NOT_FOUND, questionSubmission.getQuestion().getId()));
+
+        if(!questionSubmission.getReviews().isEmpty() || !question.getStatus().equals(Question.Status.IN_REVISION)) {
+            throw new TutorException(CANNOT_DELETE_REVIEWED_QUESTION);
+        } else {
+            removeQuestionSubmission(questionSubmission.getId());
+            question.remove();
```

Comment:
these two removes look redundant, the remove of the question should remove the submission...

Reply:
the problem is that when a question is removed, it's by a teacher, and a teacher can never delete a question that was submitted. In that case, when a student deletes a question submission and therefore calls remove() for question submission, I now remove the question there